pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+            {
+                continue;
+            }
+            pre_prepares[pre_prepare.sequence()] = this->make_signed_envelope(pre_prepare.SerializeAsString());
+        }
+
+        //  - add each preprepare created this way to the new-view message
+    }
+
+    //  - for each sequence number between the base-sequence number (the last stable checkpoint)
+    //    and the highest sequence number for which a new preprepare was created
+
+    // Fill in missing messages with no ops
+    this->fill_in_missing_pre_prepares(max_checkpoint_sequence, new_view, pre_prepares);
+
+    //  - add each preprepare created this way to the new-view message
+    // std::set<std::string> view_change_messages
+    // viewchange_messages
+
+    return this->make_newview(new_view, viewchange_envelopes_from_senders, pre_prepares);
+}
+
+bzn_envelope
+pbft::make_signed_envelope(std::string serialized_pbft_message) const
+{
+    bzn_envelope envelope;
+    envelope.set_pbft(serialized_pbft_message);
+    envelope.set_sender(this->get_uuid());
+    if (!this->crypto->sign(envelope))
+    {
+        LOG (error) << "make_signed_envelope - failed to sign envelope";
+    }
+
+    return envelope;
+}
+
+void
+pbft::save_checkpoint(const pbft_msg& msg)
+{
+    pbft_msg message;
+    for (int i{0}; i < msg.checkpoint_messages_size(); ++i)
+    {
+        bzn_envelope original_checkpoint{msg.checkpoint_messages(i)};
+
+        message.ParseFromString(original_checkpoint.pbft());
+
+        if (!this->crypto->verify(original_checkpoint))
+        {
+            continue;
+        }
+
+        this->handle_checkpoint(message, original_checkpoint);
+    }
+}
+
+void
+pbft::handle_viewchange(const pbft_msg &msg, const bzn_envelope &original_msg) {
+    if (!this->is_valid_viewchange_message(msg, original_msg))
+    {
+        LOG(error) << "handle_viewchange - invalid viewchange message, ignoring";
+        return;
+    }
+
+    this->valid_viewchange_messages_for_view[msg.view()][original_msg.sender()] = original_msg;
+
+    this->save_checkpoint(msg);
+
+    if (this->view_is_valid &&
+        this->valid_viewchange_messages_for_view[msg.view()].size() == this->max_faulty_nodes() + 1 &&
+        msg.view() > this->last_view_sent)
+    {
+        this->last_view_sent = msg.view();
+        pbft_msg viewchange_message{make_viewchange(msg.view(), this->latest_stable_checkpoint().first, this->stable_checkpoint_proof, this->prepared_operations_since_last_checkpoint())};
+        this->broadcast(this->wrap_message(viewchange_message));
+    }
+
+    // TODO move this to the end...
+    // we want to filter std::map<uint64_t, std::set<bzn_envelope>> valid_view_change_messages for
+    // 1) we would be the primary for that view
+    // 2) that view is greater than the current view
+    // 3) we have 2 f + 1 viewchange messages for it
+    const auto viewchange = std::find_if(this->valid_viewchange_messages_for_view.begin(),
```

Comment:
this comment is difficult to understand.

Reply:
That was my feeling as well. Removed

Reply:
Can we put some kind of comment on it, thought? I think the logic is bordering on "excessively clever"